File: solutions/R/stage_1/stage1_signal.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) == 0) {
  stop("Usage: Rscript stage1_signal.R <dataset_path>")
}

dataset_path <- args[[1]]
df <- read.csv(dataset_path, stringsAsFactors = FALSE)

signals <- df[df$keep == 1, ]
signals <- signals[order(signals$pos), ]

decode_digit <- function(encoded, key, pos) {
  (encoded - key - (pos * 3)) %% 10
}

digits <- mapply(decode_digit, signals$encoded, signals$key, signals$pos)
cat(paste0(digits, collapse = ""), "\n")

File: solutions/python/stage_1/stage1_signal.py
import csv
import sys

if len(sys.argv) < 2:
    raise SystemExit("Usage: python stage1_signal.py <dataset_path>")

dataset_path = sys.argv[1]
with open(dataset_path, newline="") as handle:
    rows = list(csv.DictReader(handle))

signals = sorted(
    (row for row in rows if int(row["keep"]) == 1),
    key=lambda row: int(row["pos"]),
)

digits = []
for row in signals:
    encoded = int(row["encoded"])
    key = int(row["key"])
    pos = int(row["pos"])
    digit = (encoded - key - (pos * 3)) % 10
    digits.append(str(digit))

print("".join(digits))